src/api/app/controllers/webui/badge_controller.rb

```diff
@@ -0,0 +1,31 @@
+class Webui::BadgeController < Webui::WebuiController
+  before_action :set_project
+  before_action :require_package
+
+  def show
+    results = @package.buildresult(@project, false, true).results[@package.name]
+    render inline: create_badge(results)
+  end
+
+  private
+
+  def create_badge(results)
+    filename = 'badge-unknown.svg'
+    return Rails.application.assets[filename].source if results.blank?
+
+    filename = 'badge-failed.svg' if results.any? { |r| r.code == 'failed' }
+    filename = 'badge-succeeded.svg' if results.all? { |r| r.code == 'succeeded' }
+    filename = 'badge-percent.svg' if params[:type] == 'percent'
+    prepare_badge_file(filename, results)
+  end
+
+  def prepare_badge_file(filename, results)
+    file = Rails.application.assets[filename].source
+    succeeded = results.select { |r| r.code == 'succeeded' }.try(:length).to_i
+    text = "#{100 * succeeded / results.length}%"
+    file.gsub!('%PERCENTAGE%', text)
```

Comment:
Combining the `create_badge` and `prepare_badge_file` methods would simplify the code since some code is only needed if a specific badge is needed, like this line.